### PR TITLE
Implement get_named_tzdb_transition

### LINKED ISSUE #406: get_named_tzdb_transition is unimplemented
https://github.com/boa-dev/temporal/blob/e634fe6ff90574ff74f38bee1a667bb6b867682c/src/tzdb.rs#L782

Tracking for Test262, feel free to close as a duplicate if there's somehting already open

### PR DESCRIPTION
**Pull request by Manishearth**
I found two bugs in preexisting code in the process (predating #468), hooray for writing extensive tests!:

 - When doing a tzif data search, we were returning the wrong index in one case
 - When resolving within a posix tz string, if we are on the same day as but before the start of the transition, we would return the wrong transition

Fixing those bugs has made me more confident in some of this code in general since I've reviewed it while debugging.

There's a bit of trickiness around "fake" transitions; the data can have transitions without any actual offset change; which the spec expects us to skip.

Fixes #406

### CODE REVIEW

**Comment by Manishearth** [comment]
Small issue: I still need to handle empty `transition_times` arrays here, otherwise this will crash. I'll harden the code soon.

File: src/tzdb.rs [hunk 1]
```diff
@@ -252,6 +260,163 @@ impl Tzif {
...
+                    } else {
+                        // If we found the first transition, there is no previous one,
+                        // return None
+                        return Ok(None);
+                    }
+                }
+            },
+        };
+
+        while let Some(tzif_transition) = maybe_get_transition_info(db, transition_idx) {
```
Comment (Manishearth) [inline src/tzdb.rs:320]: this code is not great and I don't like having to do it twice for last_tzif_transition

Comment (Manishearth) [reply src/tzdb.rs:320]: Thinking about this more, we actually should be calculating this later.

Comment (Manishearth) [reply src/tzdb.rs:320]: I've improved it, but I'd still love to find a testcase for some of this code

Comment (Manishearth) [reply src/tzdb.rs:320]: Some testcases:

```
Africa/Tripoli / Seconds(1382659200), 2013-10-25T02:00:00+02:00[Africa/Tripoli]
Africa/Windhoek / Seconds(1508796000), 2017-10-24T00:00:00+02:00[Africa/Windhoek]
America/Asuncion / Seconds(1728961200), 2024-10-15T00:00:00-03:00[America/Asuncion]
America/Chihuahua / Seconds(1667116800), 2022-10-30T02:00:00-06:00[America/Chihuahua]
America/Dawson / Seconds(1604214000), 2020-11-01T00:00:00-07:00[America/Dawson]
America/Dawson_Creek / Seconds(84013200), 1972-08-30T02:00:00-07:00[America/Dawson_Creek]
America/Goose_Bay / Seconds(1320116400), 2011-11-01T00:00:00-03:00[America/Goose_Bay]
America/Grand_Turk / Seconds(1520751600), 2018-03-11T03:00:00-04:00[America/Grand_Turk]
America/Indiana/Petersburg / Seconds(1194159600), 2007-11-04T02:00:00-05:00[America/Indiana/Petersburg]
America/Indiana/Vincennes / Seconds(1194159600), 2007-11-04T02:00:00-05:00[America/Indiana/Vincennes]
America/Matamoros / Seconds(1262325600), 2010-01-01T00:00:00-06:00[America/Matamoros]
America/Moncton / Seconds(1167624000), 2007-01-01T00:00:00-04:00[America/Moncton]
America/North_Dakota/Beulah / Seconds(1289116800), 2010-11-07T02:00:00-06:00[America/North_Dakota/Beulah]
America/Nuuk / Seconds(1698541200), 2023-10-28T23:00:00-02:00[America/Nuuk]
America/Ojinaga / Seconds(1669788000), 2022-11-30T00:00:00-06:00[America/Ojinaga]
America/Punta_Arenas / Seconds(1480820400), 2016-12-04T00:00:00-03:00[America/Punta_Arenas]
America/Resolute / Seconds(1173600000), 2007-03-11T03:00:00-05:00[America/Resolute]
America/Scoresbysund / Seconds(1711846800), 2024-03-31T00:00:00-01:00[America/Scoresbysund]
America/St_Johns / Seconds(1320114600), 2011-11-01T00:00:00-02:30[America/St_Johns]
America/Tijuana / Seconds(1262332800), 2010-01-01T00:00:00-08:00[America/Tijuana]
America/Whitehorse / Seconds(1604214000), 2020-11-01T00:00:00-07:00[America/Whitehorse]
Antarctica/Macquarie / Seconds(1293800400), 2011-01-01T00:00:00+11:00[Antarctica/Macquarie]
Antarctica/Palmer / Seconds(1480820400), 2016-12-04T00:00:00-03:00[Antarctica/Palmer]
Antarctica/Troll / Seconds(1108166400), 2005-02-12T00:00:00+00:00[Antarctica/Troll]
Asia/Amman / Seconds(1666908000), 2022-10-28T01:00:00+03:00[Asia/Amman]
Asia/Aqtau / Seconds(1099173600), 2004-10-31T03:00:00+05:00[Asia/Aqtau]
Asia/Atyrau / Seconds(1099173600), 2004-10-31T03:00:00+05:00[Asia/Atyrau]
Asia/Bishkek / Seconds(1123783200), 2005-08-12T00:00:00+06:00[Asia/Bishkek]
Asia/Damascus / Seconds(1666904400), 2022-10-28T00:00:00+03:00[Asia/Damascus]
Asia/Nicosia / Seconds(904597200), 1998-09-01T00:00:00+03:00[Asia/Nicosia]
Asia/Oral / Seconds(1099173600), 2004-10-31T03:00:00+05:00[Asia/Oral]
Europe/Bucharest / Seconds(852069600), 1997-01-01T00:00:00+02:00[Europe/Bucharest]
Europe/Chisinau / Seconds(852069600), 1997-01-01T00:00:00+02:00[Europe/Chisinau]
Europe/Dublin / Seconds(820454400), 1996-01-01T00:00:00+00:00[Europe/Dublin]
Europe/Guernsey / Seconds(820454400), 1996-01-01T00:00:00+00:00[Europe/Guernsey]
Europe/Isle_of_Man / Seconds(820454400), 1996-01-01T00:00:00+00:00[Europe/Isle_of_Man]
Europe/Istanbul / Seconds(1473195600), 2016-09-07T00:00:00+03:00[Europe/Istanbul]
Europe/Jersey / Seconds(820454400), 1996-01-01T00:00:00+00:00[Europe/Jersey]
Europe/Kyiv / Seconds(831934800), 1996-05-13T00:00:00+03:00[Europe/Kyiv]
Europe/Lisbon / Seconds(828234000), 1996-03-31T02:00:00+01:00[Europe/Lisbon]
Europe/London / Seconds(820454400), 1996-01-01T00:00:00+00:00[Europe/London]
Europe/Riga / Seconds(978386400), 2001-01-02T00:00:00+02:00[Europe/Riga]
Europe/Sofia / Seconds(852069600), 1997-01-01T00:00:00+02:00[Europe/Sofia]
Europe/Tallinn / Seconds(1014242400), 2002-02-21T00:00:00+02:00[Europe/Tallinn]
Europe/Vilnius / Seconds(1041372000), 2003-01-01T00:00:00+02:00[Europe/Vilnius]
Pacific/Guam / Seconds(977493600), 2000-12-23T00:00:00+10:00[Pacific/Guam]
Pacific/Norfolk / Seconds(1561899600), 2019-07-01T00:00:00+11:00[Pacific/Norfolk]
Pacific/Saipan / Seconds(977493600), 2000-12-23T00:00:00+10:00[Pacific/Saipan]
```

Comment (nekevss) [reply src/tzdb.rs:320]: As a late heads up, I have a handful of edge cases in the zoneinfo test file under the `zoneinfo` directory. Almost all of those (outside of one or two) were added because they represented parsing / compiling edge cases I came across. Although, this appears to be a more comprehensive list to be honest.

File: src/builtins/compiled/zoneddatetime.rs [hunk 2]
```diff
@@ -577,4 +579,179 @@ mod tests {
...
+                .unwrap()
+                .to_string(),
+            s
+        );
+    }
+
+    // Modern dates
+
+    // Transitions
+    const DST_2025_03_09: &str = "2025-03-09T03:00:00-07:00[America/Los_Angeles]";
```
Comment (nekevss) [inline src/builtins/compiled/zoneddatetime.rs:597]: praise: great tests!

File: src/builtins/core/zoneddatetime.rs [hunk 3]
```diff
@@ -803,12 +803,21 @@ impl ZonedDateTime {
...
+        };
 
-        Ok(result)
+        if transition.check_validity().is_err() {
+            // GetNamedTimeZoneNextTransition, GetNamedTimeZonePreviousTransition include a check for out-of-bounds
+            // instants. Instead of requiring providers handle that, we handle it here.
+            return Ok(None);
+        }
+        // 12. Return ! CreateTemporalZonedDateTime(transition, timeZone, zonedDateTime.[[Calendar]]).
+        Ok(Some(ZonedDateTime::try_new(
```
Comment (nekevss) [inline src/builtins/core/zoneddatetime.rs]: nit: the spec has this call as infallible.

File: src/tzdb.rs [hunk 4]
```diff
@@ -252,6 +260,177 @@ impl Tzif {
...
     }
 
+    pub fn get_named_tz_transition(
+        &self,
+        epoch_nanoseconds: i128,
+        direction: TransitionDirection,
+    ) -> TemporalResult<Option<EpochNanoseconds>> {
+        // First search the tzif data
+
+        let epoch_seconds = Seconds((epoch_nanoseconds / 1_000_000_000) as i64);
```
Comment (nekevss) [inline src/tzdb.rs:270]: thought(out-of-scope): It could be nice to use `Seconds` instead of nanoseconds or some other type than `i128`

This value could store the seconds value and an `is_exact` flag and have `from_seconds`, `from_milliseconds`, and `from_nanoseconds`, where the first two steps here -- the div_mod -- are part of the types ctors. It could help generalize the trait in comparison to i128.

**Review by nekevss — APPROVED** [review]
Looks good to me overall. Have one nit with a specification discrepancy that would be nice to have cleaned up to make #245 a bit easier.

File: src/builtins/core/zoneddatetime.rs (hunk 3 above)
Comment (Manishearth) [reply src/builtins/core/zoneddatetime.rs]: I kept a temporal_unwrap() around; I'd rather not have release builds panicking because we missed something

File: src/tzdb.rs (hunk 4 above)
Comment (Manishearth) [reply src/tzdb.rs:270]: I don't think that would work, `is_exact` is messy to reason about in a generic context. If we want to we should just use ns here.
